### ISSUE TITLE
Export throwPluginError function

### ISSUE THREAD

**Issue by vladshcherbin** [rollup/rollup#3159]
### Feature Use Case

When creating plugins, it's a common thing to check provided options for required ones and when they are missing, throw an error:

```js
export default function myPlugin(options = {}) {
  const { requiredOption } = options

  if (!requiredOption) {
    throw new Error('"requiredOption" is a required option')
  }
}
```

Since it's thrown not in a hook, it will have this format:

```console
Error: "requiredOption" is a required option
```

If an error is throw in a hook, it will take plugin name and error will be:

```console
(plugin my-plugin) Error: "requiredOption" is a required option
```

Even if I change error message (thrown not in a hook) to have plugin name, it will still be different:

```console
Error: (plugin my-plugin) "requiredOption" is a required option
``` 

### Feature Proposal

It would be great to have exported `throwPluginError` function which can be simply used with passed plugin name and message to unify all plugin errors.

**Comment by lukastaegert** [comment]
Hi, I see your point but I am very opposed to the pattern of plugins having a dependency on Rollup as this is causing all kinds of dependency issues. Acorn’s decision to do just that for instance forced us to make acorn an external dependency and not have acorn plugins in the browser build. My recommendation would be to defer reporting the error to the buildStart hook. This may be slightly more awkward to set up but will work today and will provide you with the `this.error` context function.

**Comment by lukastaegert** [comment]
Or to rephrase the problem, there is no guarantee that the Rollup version which the plugin is importing the handler from is actually the Rollup version that is running, and it will also tie your plugin to Node even if it may not be necessary.

**Comment by vladshcherbin** [comment]
@lukastaegert hm, I might not fully understand the problem.

I can see 2 entries in `src` - browser and node, both of can use plugins and have `throwPluginError` function inside. So, the error should work the same in both entries 🤔

Version check for this addition can be made through `peerDependency`, how it's done for example in [rollup-plugin-node-resolve](https://github.com/rollup/rollup-plugin-node-resolve/blob/master/package.json#L44) and if the plugin supports browser usage, version can also be checked with `VERSION` export.

Alternatively, this function may be exposed only to node plugins with a note in the docs if there are any issues with browser entry.

Correct me if I'm wrong.

**Comment by vladshcherbin** [comment]
I also thought about using a hook for that, checked multiple plugins and found same thing I do - outside hooks check. To me, using hook for this feels like an illogical workaround or limitation so I skipped this idea and decided to open a feature request :)

**Comment by vladshcherbin** [comment]
I also found same limitation in `rollup-plugin-node-resolve`, for example:

- [here](https://github.com/rollup/rollup-plugin-node-resolve/blob/v5.2.0/src/index.js#L54) the error message starts with `node-resolve: do not use deprecated ...`

- later, [1](https://github.com/rollup/rollup-plugin-node-resolve/blob/v5.2.0/src/index.js#L127) / [2](https://github.com/rollup/rollup-plugin-node-resolve/blob/v5.2.0/src/index.js#L200) errors are inconsistent, some are thrown with plugin name, some are not, some with `new Error`, some with `this.error`.

So, I believe already existing plugins will also benefit from this feature because plugin errors are not uniformed and if you have multiple plugins, sometimes you can't even understand where the error is coming from.

**Comment by lukastaegert** [comment]
I may be biased from our bad experience with acorn, but plugins importing the plugin system, in this case Rollup, is a slippery slope and an architecture anti-pattern that I will definitely prevent if possible. Plugins should be given access to the system which they are plugins for solely by means of dependency injection, though the exact mechanism is certainly to be discussed. This restriction will ensure portability of the system and is not discussable.

The fact that existing plugins do not do it the way I suggested is because their logic predates the existence of those plugins hooks and nobody put work into changing them. As such, they are bad examples to copy from, PR definitely welcome. At the moment there are far too few people doing work on these plugins, so please excuse that they do not follow the current best practices, it is sad as it is.

The way Rollup plugins are instantiated unfortunately means they do not have access to the Rollup instance during their instantiation, which means that I see hooks as the only option to properly throw errors if you want them associated with the plugin.

**Comment by lukastaegert** [comment]
If this is just about a common way to format error messages, rollup-pluginutils could be a place to put such a helper, though.

**Comment by lukastaegert** [comment]
Actually it would be THE proper place.

**Comment by vladshcherbin** [comment]
@lukastaegert yes, the issue I have is about having same format in error messages (so that error format outside hooks doesn't differ from inside hooks) and if it will be possible to achieve this with using `rollup-pluginutils` helper, it would be totally fine. 👌

I can also help with PRs in other plugins to help with errors consistency after helper release.

**Comment by vladshcherbin** [comment]
Another thing to think about hook which I also tried, `buildStart` for example is async parallel and some plugin might work before error is thrown in another plugin. If error is thrown outside hook, I believe it will be faster, build won't start and you won't have this parallel plugin condition. If I'm not mistaken :)

**Comment by lukastaegert** [comment]
I guess you are right.

So how about we add a utility function to `rollup-pluginutils` e.g. `throwInitializationError`? The documentation should explain that this is for consistent output of errors during the creation of the plugin when no Rollup instance is available, but that in hooks (other than `options` at the moment which has no proper context), `this.error` should be used.

A PR, even an incomplete one, would be very much appreciated to get this started in `rollup-pluginutils`.

**Comment by vladshcherbin** [comment]
@lukastaegert hey, I'm wondering how this error function should work? [Here](https://github.com/rollup/rollup/blob/v1.25.0/cli/logging.ts#L11) I found it's just a `console.error` call with `process.exit` call later.

Should function in `rollup-pluginutils` do the same? (Regular error throw won't work as currently plugin name comes before Error word)

**Comment by lukastaegert** [comment]
I think it  makes sense as configuration errors are not "fixable". The beauty is that we can still change it later if there are valid arguments against a `process.exit`. So yes, go for that approach.
